zone/bot_raid.cpp

```diff
@@ -0,0 +1,287 @@
+				if (m.member->GetHPRatio() <= hpr) {
+					need_healed++;
+				}
+
+				if (includePets) {
+					if (m.member->GetPet() && m.member->GetPet()->GetHPRatio() <= hpr) {
+						need_healed++;
+					}
+				}
+			}
+		}
+	}
+	return need_healed;
+}
+
+void Bot::ProcessRaidInvite(Mob* invitee, Client* invitor, bool group_invite) {
+
+	if (!invitee || !invitor) {
+		return;
+	}
+
+	if (invitee->IsBot() &&
+		invitor->HasRaid() &&
+		invitee->GetOwner()->HasRaid() &&
+		invitor->GetRaid()->IsRaidMember(invitee->GetOwner()->GetName())
+	) {
+		// If the Bot Owner is in our raid we need to be able to invite their Bots
+	}
+	else if (invitee->IsBot() && (invitee->CastToBot()->GetBotOwnerCharacterID() != invitor->CharacterID())) {
+		invitor->Message(
+			Chat::Red,
+			fmt::format(
+				"{} is not your Bot. You can only invite your own Bots, or Bots that belong to a Raid member.",
+				invitee->GetCleanName()
+			).c_str()
+		);
+		return;
+	}
+
+	Raid* raid = entity_list.GetRaidByClient(invitor);
+
+	Bot::CreateBotRaid(invitee, invitor, group_invite, raid);
+}
+
+void Bot::CreateBotRaid(Mob* invitee, Client* invitor, bool group_invite, Raid* raid) {
+
+	Group* g_invitee = invitee->GetGroup();
+	Group* g_invitor = invitor->GetGroup();
+
+	if (g_invitee && invitor->IsClient()) {
+		if (!g_invitee->IsLeader(invitee)) {
+			invitor->Message(
+				Chat::Red,
+				fmt::format(
+					"You can only invite group leaders or ungrouped bots. Try {} instead.",
+					g_invitee->GetLeader()->GetCleanName()
+				).c_str()
+			);
+			return;
+		}
+	}
+
+	bool new_raid = false;
+	if (!raid) {
+		new_raid = true;
+		raid = new Raid(invitor);
+		entity_list.AddRaid(raid);
+		raid->SetRaidDetails();
+	}
+
+	// Add Invitor to new raid
+	if (new_raid) {
+		if (g_invitor) {
+			ProcessBotGroupAdd(g_invitor, raid, nullptr, true, true);
+		} else {
+			raid->SendRaidCreate(invitor);
+			raid->AddMember(invitor, 0xFFFFFFFF, true, false, true);
+			raid->SendMakeLeaderPacketTo(invitor->GetName(), invitor);
+			if (raid->IsLocked()) {
+				raid->SendRaidLockTo(invitor);
+			}
+		}
+	}
+
+	// Add Bot Group or Client Bot Group to raid
+	if (g_invitee) {
+		ProcessBotGroupAdd(g_invitee, raid);
+
+	// Add individual client to raid
+	} else if (invitee->IsClient()) {
+		ProcessBotGroupAdd(g_invitee, raid, invitee->CastToClient());
+
+	// Add individual bot to raid
+	} else {
+		auto gid = raid->GetGroup(invitor->GetName());
+		auto b = invitee->CastToBot();
+
+		// gives us a choice to either invite directly into the clients Raid Group, or just into the Raid
+		if (group_invite && raid->GroupCount(gid) < MAX_GROUP_MEMBERS) {
+			raid->AddBot(b, gid);
+		} else {
+			raid->AddBot(b);
+		}
+
+		if (new_raid) {
+			invitee->SetFollowID(invitor->GetID());
+		}
+	}
+}
+
+void Bot::ProcessBotGroupAdd(Group* group, Raid* raid, Client* client, bool new_raid, bool initial) {
+
+	uint32 raid_free_group_id = raid->GetFreeGroup();
+	if (group) {
+		for (int x = 0; x < MAX_GROUP_MEMBERS; x++) {
```

Comment:
You can range through the array versus doing the integer iteration

eg

```
for (auto & m : members) {
```

Accessing

```cpp
m.member->DoThing()
```